### PR TITLE
[INTEL MKL] scatter optimizations.

### PR DESCRIPTION
**Pull request by Srini511**
Parallelizing scatter update op.

### CODE REVIEW

File: tensorflow/core/kernels/scatter_functor.h [hunk 1]
```diff
@@ -217,6 +258,30 @@ struct ScatterFunctorBase {
     }
     return -1;
   }
+
+  Index operator()(OpKernelContext* c, const Device& d,
+                   typename TTypes<T>::Matrix params,
+                   typename TTypes<T>::ConstMatrix updates,
+                   typename TTypes<Index>::ConstFlat indices) {
+    // indices and params sizes were validated in DoCompute().
```
Comment (penpornk) [inline tensorflow/core/kernels/scatter_functor.h]: I tried this PR internally, and it's still significantly slower than the serial version. (~15% slower for small cases, up to 16x slower for large cases.) I still don't have time to look at it. So sorry about that! To not block you further, Let's use the parallel code only when `PLATFORM_GOOGLE` is not defined. Could you please wrap the code with `#ifdef PLATFORM_GOOGLE`?
```
#ifdef PLATFORM_GOOGLE
    // The parallel version is significantly slower internally. Only call the
    // serial version for now.
    // TODO(penporn): Avoid locking in parallelization (sort beforehand).
    return SerialExecute(c, d, params, updates, indices);
#else
     // indices and params sizes were validated in DoCompute().
    const Index N = static_cast<Index>(indices.size());
    const Index limit = static_cast<Index>(params.dimension(0));
    ...
#endif  // PLATFORM_GOOGLE
```

File: tensorflow/core/kernels/scatter_functor.h [hunk 2]
```diff
@@ -217,6 +258,30 @@ struct ScatterFunctorBase {
...
+    const Index ser_par_ratio = 10000;
+    // For parallelizing the updates, duplicate entries need to be handled
+    // correctly.Multiple updates to the same index has to be serialized.
+    // This can lead to lock contention which may nullify the benefits of
+    // parallelization. Assuming uniform random distribution of the indices, we
+    // come up with a rough heuristic and determine whether the updates execute
+    // serially or parallelly. Also if 'N' is small, overheads of parallel
+    // execution outweigh its benefits and hence we check the value of N.
+    const bool execute_serial =
+        ((N < min_n_threshold) || ((N / limit) > ser_par_ratio)) ? true : false;
```
Comment (penpornk) [inline tensorflow/core/kernels/scatter_functor.h]: The expression already returns a boolean. We can drop the ternary operator.
```suggestion
        ((N < min_n_threshold) || ((N / limit) > ser_par_ratio));
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for re-submitting the PR! I'm sorry I haven't had time to resolve the internal issue yet.

Note: This is bringing back https://github.com/tensorflow/tensorflow/pull/30880.

File: tensorflow/core/kernels/scatter_functor.h (hunk 1 above)
Comment (Srini511) [reply tensorflow/core/kernels/scatter_functor.h]: Hi @penpornk , Are you running the benchmarks in scatter_op_test? I  ran the test, with the PR and without the PR using clang and this is what i see.
With PR
BM_ScatterAddInt32/1               3112     232906       321.4M items/s
BM_ScatterAddInt32/10              5009     100000       1996.4M items/s
BM_ScatterAddInt32/64             14396      45221       4445.6M items/s
BM_ScatterAddInt32/256            94204       7372       2717.5M items/s
BM_ScatterAddInt32/1024          327392       2186       3127.7M items/s
BM_ScatterAddInt32Large/1       7622000        100       131.2M items/s
BM_ScatterAddInt32Large/10      7225810        100       1383.9M items/s
BM_ScatterAddInt32Large/64     10714610        100       5973.2M items/s
BM_ScatterAddInt32Large/256    38700000        100       6615.0M items/s
BM_ScatterAddInt32Large/1024  105354390        100       9719.6M items/s

No PR (serial execute always)

BM_ScatterAddInt32/1               3132     222999       319.2M items/s
BM_ScatterAddInt32/10              5003     141062       1998.7M items/s
BM_ScatterAddInt32/64             13886      50565       4609.1M items/s
BM_ScatterAddInt32/256            96351       7565       2657.0M items/s
BM_ScatterAddInt32/1024          327748       2168       3124.4M items/s
BM_ScatterAddInt32Large/1       9867540        100       101.3M items/s
BM_ScatterAddInt32Large/10     20585780        100       485.8M items/s
BM_ScatterAddInt32Large/64     51982330        100       1231.2M items/s
BM_ScatterAddInt32Large/256   165220150        100       1549.4M items/s
BM_ScatterAddInt32Large/1024  602341830        100       1700.0M items/s

I used this command line. 
bazel run --config=download_clang -c opt //tensorflow/core/kernels:scatter_op_test -- --benchmarks=..
Please let me know if anything else is missed.

Comment (Srini511) [reply tensorflow/core/kernels/scatter_functor.h]: we were also thinking along the lines of sorting to improve memory locality but did not evaluate that option.

**Comment by Srini511** [comment]
@penpornk made all the changes requested except defaulting to Serial Execute for google internal runs. Will wait for your comments for that.

File: tensorflow/core/kernels/scatter_functor.h (hunk 1 above)
Comment (penpornk) [reply tensorflow/core/kernels/scatter_functor.h]: Thank you for testing with clang! Yes, I see similar speedups with external TF with clang. The problem with the internal one (i.e., same TF code built with internal toolchain) is with threading. 
Serial scatter: Internal consistently faster than external (clang) by a constant factor.
Parallel scatter: Internal is so much slower than its external counterpart or even the internal serial scatter (up to 16x slower than serial).

I'm not even sure if this is the locking overhead. Might just be threading overhead too. (We link to a different threading library internally.) So I don't think you'd be able to replicate the issue. :( I don't have time to look into this yet (not in the next two weeks). If you'd like to get it in soon then the best way is to just wrap it with `PLATFORM_GOOGLE`. This way normal TensorFlow (external) will benefit from your optimization too.

**Review by penpornk — COMMENTED** [review]
Thank you for the quick fixes!

**Comment by Srini511** [comment]
Hi @penpornk Thanks for the response. I have wrapped the code around ifdef

**Comment by Srini511** [comment]
@penpornk gentle reminder to review the latest change

**Review by penpornk — APPROVED** [review]
Thank you so much for your patience!

File: tensorflow/core/kernels/scatter_op_test.cc [hunk 3]
```diff
@@ -175,6 +186,38 @@ TEST_F(ScatterUpdateOpTest, Error_IndexOutOfRange) {
...
+  // Feed and run
+  const int kRows = 1;
+  std::vector<int32> values(kRows, 0);
+  const int kNumUpdates = 1000000;
+  std::vector<int32> indices(kNumUpdates, 0);
+  std::vector<int32> updates(kNumUpdates, 1);
+  AddInputFromArray<int32>(TensorShape({kRows}), values);
+  AddInputFromArray<int32>(TensorShape({kNumUpdates}), indices);
+  AddInputFromArray<int32>(TensorShape({kNumUpdates}), updates);
+  testing::ItemsProcessed((static_cast<int64>(kNumUpdates)));
```
Comment (penpornk) [inline tensorflow/core/kernels/scatter_op_test.cc]: This can only be called from a benchmark and not a test. Please remove this line.

**Review by penpornk — CHANGES_REQUESTED** [review]
We got a test failure. Could you please help fix this? Thank you!

**Comment by Srini511** [comment]
Hi @penpornk made the changes. Kindly review.

**Review by penpornk — APPROVED** [review]
Thank you very much!

**Comment by Srini511** [comment]
HI @penpornk Just checking on the status of this PR

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.

**Pull request closed without merging**